modules/service/src/main/resources/db/migration/V0847__observation_reference.sql

```diff
@@ -0,0 +1,92 @@
+-- Add an index column to the observation table.
+ALTER TABLE t_observation
+  ADD COLUMN c_observation_index int4 CHECK (c_observation_index > 0),
+  ADD CONSTRAINT unique_observation_index_per_program UNIQUE (c_program_id, c_observation_index);
+
+-- A function to obtain the next observation number for a particular program.
+CREATE OR REPLACE FUNCTION next_observation_index(pid d_program_id)
+RETURNS INT
+AS $$
+DECLARE
+    sequence_name VARCHAR;
+BEGIN
+    sequence_name := CONCAT('s_', REPLACE(pid::text, '-', '')); -- a hyphen (as in p-123) will not work
+
+    BEGIN
+        EXECUTE 'CREATE SEQUENCE IF NOT EXISTS ' || sequence_name;
+    EXCEPTION WHEN unique_violation THEN
+        NULL; -- sequence exists, this was an attempt to create it in parallel
+    END;
```

Comment:
I did this trick of creating a sequence on the fly elsewhere.  I pulled it out again for maintaining the observation index for each program.  Unfortunately it produced unique constraint violations in one particular unrelated test case.

```
skunk.exception.PostgresErrorException: 
🔥  
🔥  Postgres ERROR 23505 raised in _bt_check_unique (nbtinsert.c:664)
🔥  
🔥    Problem: Duplicate key value violates unique constraint
🔥             "pg_class_relname_nsp_index".
🔥     Detail: Key (relname, relnamespace)=(s_p101, 2200) already exists.
```

The test case was in `lucuma.odb.graphql.query.constraintSetGroup` where we do:

```
createObservation(user, pid, iq, sb).replicateA(2)
```

which I assume is somehow creating the same sequence in parallel.  Adding this exception handler "fixes" the problem but I wonder if it discredits the entire approach.  I'm not sure what else to do though other than reserving a sequence when the program is created I suppose.

Reply:
I could do that pretty easily in this case.  For the semester counters I could just write a mutation that reserves them for the foreseeable future.

Reply:
I think maybe go with the exception handler. It's weird but it does seem to fix the issue?